### PR TITLE
[tinyfiledialogs] 3.6.3 -> 3.8.8

### LINKED ISSUE #17929: [tinyfiledialogs] update to 3.8.8
Library name: tinyfiledialogs

New version number: 3.8.8

official link:
https://sourceforge.net/projects/tinyfiledialogs/

### PR DESCRIPTION
**Pull request by SamuelMarks**
**Describe the pull request**
https://sourceforge.net/p/tinyfiledialogs/code/ci/ab6f4f916aaa95d05247ffa66a30867e7f55e875/tree/

He didn't release a new zip so gotta change from `vcpkg_from_sourceforge` [which lists this package as an example] with `vcpkg_download_distfile`… not sure if there's a better way. Maybe it needs to be actually cloned rather than a snapshot downloaded? - Or maybe it needs to follow redirects? (like `curl`'s `-L` flag).

EDIT: Switched to `vcpkg_from_git`; which seems to make the whole thing work.

- #### What does your PR fix?  
Upgrades the version

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  <all / linux, windows, ...>, <Yes/No>

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  `Your answer`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

---
WiP, test with:

### `CMakeLists.txt`
```cmake
cmake_minimum_required(VERSION 3.4)
cmake_policy(SET CMP0048 NEW)
project(pp VERSION 0.0.0 LANGUAGES C)

file(DOWNLOAD "https://sourceforge.net/p/tinyfiledialogs/code/ci/4b65621952f2cf1045837a6380aa912b028daa47/tree/hello.c?format=raw" "src/main.c"
     EXPECTED_HASH SHA256=72502b4b5e93bf262b1ff148de028d12a187c1fda45de89212a3f896e2be1827)

set(src "${CMAKE_CURRENT_BINARY_DIR}/src/main.c")

find_package(tinyfiledialogs CONFIG REQUIRED)

add_executable("${PROJECT_NAME}" "${src}")

target_link_libraries("${PROJECT_NAME}" PRIVATE tinyfiledialogs::tinyfiledialogs)
```

EDIT: I was able to get it to work like so, but the help text doesn't tell the user to do this:
```cmake
cmake_minimum_required(VERSION 3.4)
cmake_policy(SET CMP0048 NEW)
project(pp VERSION 0.0.0 LANGUAGES C)

file(DOWNLOAD "https://sourceforge.net/p/tinyfiledialogs/code/ci/4b65621952f2cf1045837a6380aa912b028daa47/tree/hello.c?format=raw" "src/main.c"
     EXPECTED_HASH SHA256=72502b4b5e93bf262b1ff148de028d12a187c1fda45de89212a3f896e2be1827)

set(src "${CMAKE_CURRENT_BINARY_DIR}/src/main.c")

find_package(tinyfiledialogs CONFIG REQUIRED)
find_path(TINYFILEDIALOGS_INCLUDE_DIRS "tinyfiledialogs.h")

add_executable("${PROJECT_NAME}" "${src}")
target_include_directories("${PROJECT_NAME}" PRIVATE "${TINYFILEDIALOGS_INCLUDE_DIRS}")
target_link_libraries("${PROJECT_NAME}" PRIVATE tinyfiledialogs::tinyfiledialogs)
```

### CODE REVIEW

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
@SamuelMarks , thaks for your contribution

**Comment by SamuelMarks** [comment]
@JonLiu1993 No worries.

On a related note I tried to make this header-only, by removing the `CMakeLists.txt` and changing `portfile.cmake` to:
```cmake
vcpkg_fail_port_install(ON_TARGET "uwp")

set(TINYFILEDIALOGS_VERSION "3.8.7")
set(SOURCE_PATH "${CURRENT_BUILDTREES_DIR}/src/tinyfiledialogs-${TINYFILEDIALOGS_VERSION}")

vcpkg_from_git(
    OUT_SOURCE_PATH "${SOURCE_PATH}"
    URL "https://git.code.sf.net/p/tinyfiledialogs/code"
    REF "ab6f4f916aaa95d05247ffa66a30867e7f55e875"
)
message("SOURCE_PATH = ${SOURCE_PATH}")
file(COPY "${${SOURCE_PATH}}/tinyfiledialogs.h" "${${SOURCE_PATH}}/tinyfiledialogs.c" DESTINATION "${SOURCE_PATH}")

# tinyfiledialogs.amalgamation.h
file(READ "${SOURCE_PATH}/tinyfiledialogs.c" _c)
file(READ "${SOURCE_PATH}/tinyfiledialogs.h" _h)
string(FIND "${_c}" "char tinyfd_version" _c_source_idx)
string(SUBSTRING "${_c}" 0 "${_c_source_idx}" _c_source_head)
string(SUBSTRING "${_c}" "${_c_source_idx}" -1 _c_source)
file(WRITE "${SOURCE_PATH}/tinyfiledialogs.h" "${_h}" "${_c_source_head}" "${_c_source}")

file(INSTALL "${SOURCE_PATH}/tinyfiledialogs.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
file(READ "${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs.h" _contents)
string(SUBSTRING "${_contents}" 0 1024 _contents)
file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright" "${_contents}")

vcpkg_copy_pdbs()
```

But currently getting a bunch too many `error: redefinition of` issues. Any suggestions?

**Comment by JackBoosY** [comment]
@SamuelMarks  What problem did you encounter?

**Comment by SamuelMarks** [comment]
So the code in this PR works. It just requires a library linkage. I'm trying to get it working without that extra step by combining the .c and .h file into one, but that gives me these errors:
```
tinyfiledialogs.h:621:6:  error: redefinition of 'tfd_replaceSubStr'
tinyfiledialogs.h:598:13: error: redefinition of 'RGB2Hex'
tinyfiledialogs.h:574:13: error: redefinition of 'Hex2RGB'
tinyfiledialogs.h:561:13: error: redefinition of 'ensureFinalSlash'
tinyfiledialogs.h:531:15: error: redefinition of 'getLastName'
tinyfiledialogs.h:501:15: error: redefinition of 'getPathWithoutFinalSlash'
```

Is there a trick to this? - Like maybe walking the symbols ensuring the prototypes appear at the top and the implementations at the bottom, and no duplication?

**Comment by SamuelMarks** [comment]
@JackBoosY As it stands however, this can definitely be merged into master.

**Comment by JackBoosY** [comment]
> @JonLiu1993 No worries.
> 
> On a related note I tried to make this header-only, by removing the `CMakeLists.txt` and changing `portfile.cmake` to:
> 
> ```cmake
> vcpkg_fail_port_install(ON_TARGET "uwp")
> 
> set(TINYFILEDIALOGS_VERSION "3.8.7")
> set(SOURCE_PATH "${CURRENT_BUILDTREES_DIR}/src/tinyfiledialogs-${TINYFILEDIALOGS_VERSION}")
> 
> vcpkg_from_git(
>     OUT_SOURCE_PATH "${SOURCE_PATH}"
>     URL "https://git.code.sf.net/p/tinyfiledialogs/code"
>     REF "ab6f4f916aaa95d05247ffa66a30867e7f55e875"
> )
> 
> # tinyfiledialogs.amalgamation.h
> file(READ "${${SOURCE_PATH}}/tinyfiledialogs.c" _c)
> file(READ "${${SOURCE_PATH}}/tinyfiledialogs.h" _h)
> string(FIND "${_c}" "char tinyfd_version" _c_source_idx)
> string(SUBSTRING "${_c}" 0 "${_c_source_idx}" _c_source_head)
> string(SUBSTRING "${_c}" "${_c_source_idx}" -1 _c_source)
> file(WRITE "${SOURCE_PATH}/tinyfiledialogs.h" "${_h}" "${_c_source_head}" "${_c_source}")
> 
> file(INSTALL "${SOURCE_PATH}/tinyfiledialogs.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
> file(READ "${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs.h" _contents)
> string(SUBSTRING "${_contents}" 0 1024 _contents)
> file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright" "${_contents}")
> 
> vcpkg_copy_pdbs()
> ```
> 
> But currently getting a bunch too many `error: redefinition of` issues. Any suggestions?

I want to know why you need to do this, can you explain more?

**Comment by SamuelMarks** [comment]
There's no real reason for it. I just think that seeing as there are no external dependencies it would be nice if this library were header-only. That way it's simpler to manage its dependency.

Also `vcpkg` isn't telling the user to include the `find_path` and `target_include_directories` lines.

**Review by JackBoosY — APPROVED** [review]

File: ports/tinyfiledialogs/portfile.cmake [hunk 1]
```diff
@@ -2,25 +2,28 @@ vcpkg_fail_port_install(ON_TARGET "uwp")
...
 vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
 
-vcpkg_from_sourceforge(
-    OUT_SOURCE_PATH SOURCE_PATH
-    REPO tinyfiledialogs
-    FILENAME "frozen_versions/tinyfiledialogs-3.6.3.zip"
-    SHA512 42c3bd34b0287cf2477f9ede049bea29a9306304e8fab7740065957d3737f4041899f26f29a0693e801cb0a7b63844509f86441262303ff0a4030a431ffac648
+set(TINYFILEDIALOGS_VERSION "3.8.7")
+set(SOURCE_PATH "${CURRENT_BUILDTREES_DIR}/src/tinyfiledialogs-${TINYFILEDIALOGS_VERSION}")
+
```
Comment (ras0219-msft) [inline ports/tinyfiledialogs/portfile.cmake:7]:
```suggestion
```

File: ports/tinyfiledialogs/portfile.cmake [hunk 2]
```diff
@@ -2,25 +2,28 @@ vcpkg_fail_port_install(ON_TARGET "uwp")
...
+vcpkg_from_git(
+    OUT_SOURCE_PATH "${SOURCE_PATH}"
```
Comment (ras0219-msft) [inline ports/tinyfiledialogs/portfile.cmake:9]:
```suggestion
    OUT_SOURCE_PATH SOURCE_PATH
```
This parameter is a variable name that will be set to the location of the sources (an "output variable")

File: ports/tinyfiledialogs/portfile.cmake [hunk 3]
```diff
@@ -2,25 +2,28 @@ vcpkg_fail_port_install(ON_TARGET "uwp")
...
+    URL "https://git.code.sf.net/p/tinyfiledialogs/code"
+    REF "ab6f4f916aaa95d05247ffa66a30867e7f55e875"
 )
 
-file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})
+file(COPY "${${SOURCE_PATH}}/tinyfiledialogs.h" "${${SOURCE_PATH}}/tinyfiledialogs.c" DESTINATION "${SOURCE_PATH}")
```
Comment (ras0219-msft) [inline ports/tinyfiledialogs/portfile.cmake:14]:
```suggestion
```

File: ports/tinyfiledialogs/portfile.cmake [hunk 4]
```diff
@@ -2,25 +2,28 @@ vcpkg_fail_port_install(ON_TARGET "uwp")
...
 vcpkg_fixup_cmake_targets()
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
-
-file(READ ${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs/tinyfiledialogs.h _contents)
+file(INSTALL "${SOURCE_PATH}/tinyfiledialogs.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
+file(READ "${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs/tinyfiledialogs.h" _contents)
 string(SUBSTRING "${_contents}" 0 1024 _contents)
-file(WRITE ${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright "${_contents}")
+file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright" "${_contents}")
+file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include" "${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs")
```
Comment (ras0219-msft) [inline ports/tinyfiledialogs/portfile.cmake:29]:
```suggestion
file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
```
This deletes the include files; have you made sure that the resulting package works?

**Review by ras0219-msft — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: ports/tinyfiledialogs/portfile.cmake [hunk 5]
```diff
@@ -2,25 +2,28 @@ vcpkg_fail_port_install(ON_TARGET "uwp")
...
     PREFER_NINJA
 )
 
 vcpkg_install_cmake()
 
 vcpkg_fixup_cmake_targets()
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
-
-file(READ ${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs/tinyfiledialogs.h _contents)
+file(INSTALL "${SOURCE_PATH}/tinyfiledialogs.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
```
Comment (ras0219-msft) [inline ports/tinyfiledialogs/portfile.cmake:25]:
```suggestion
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]
Please run command `vcpkg x-add-version --all` then commit the changes.

**Comment by BillyONeal** [comment]
Please merge with master before requesting a build again to get the current MacOS build fleet

File: versions/t-/tinyfiledialogs.json [hunk 6]
```diff
@@ -1,5 +1,15 @@
...
+    {
+      "git-tree": "1c6a12747cdc2688cd51009f29c282ec5ad7d465",
+      "version-string": "3.8.8",
+      "port-version": 0
+    },
+    {
+      "git-tree": "12edbf7a3f4fdf7bf050c9a22bf25fd06a2a7922",
+      "version-string": "3.8.7",
+      "port-version": 0
+    },
```
Comment (JackBoosY) [inline versions/t-/tinyfiledialogs.json:12]:
```suggestion
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Review by JackBoosY — APPROVED** [review]

File: ports/tinyfiledialogs/portfile.cmake [hunk 7]
```diff
@@ -2,25 +2,23 @@ vcpkg_fail_port_install(ON_TARGET "uwp")
...
 )
 
 vcpkg_install_cmake()
 
 vcpkg_fixup_cmake_targets()
-file(REMOVE_RECURSE ${CURRENT_PACKAGES_DIR}/debug/include)
-
-file(READ ${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs/tinyfiledialogs.h _contents)
+file(READ "${CURRENT_PACKAGES_DIR}/include/tinyfiledialogs/tinyfiledialogs.h" _contents)
 string(SUBSTRING "${_contents}" 0 1024 _contents)
```
Comment (strega-nil-ms) [inline ports/tinyfiledialogs/portfile.cmake:22]:
```suggestion
# reads between the line "- License -" and a closing "*/"
if (NOT _contents MATCHES [[- License -(([^*]|\*[^/])*)\*/]])
	message(FATAL_ERROR "Failed to parse license from tinyfiledialogs.h")
endif()
```

File: ports/tinyfiledialogs/portfile.cmake [hunk 8]
```diff
@@ -2,25 +2,23 @@ vcpkg_fail_port_install(ON_TARGET "uwp")
...
-file(WRITE ${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright "${_contents}")
+file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright" "${_contents}")
```
Comment (strega-nil-ms) [inline ports/tinyfiledialogs/portfile.cmake:23]:
```suggestion
file(WRITE "${CURRENT_PACKAGES_DIR}/share/${PORT}/copyright" "${CMAKE_MATCH_1}")
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]

**Comment by strega-nil-ms** [comment]
@SamuelMarks I can't push to your branch, so would you mind making the changes I suggested? I know they're unrelated to your PR but might as well while you're making changes :)

File: versions/t-/tinyfiledialogs.json [hunk 9]
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "1c6a12747cdc2688cd51009f29c282ec5ad7d465",
```
Comment (JackBoosY) [inline versions/t-/tinyfiledialogs.json:4]:
```suggestion
      "git-tree": "1a5b599e8303e1b62bca446461bd7114ec97caba",
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

**Comment by xarthurx** [comment]
For `3.8.8`, the author did provide a new `zip`. So perhaps using the original workflow to directly update to the newest would be possible?

**Comment by JonLiu1993** [comment]
> For `3.8.8`, the author did provide a new `zip`. So perhaps using the original workflow to directly update to the newest would be possible?

Good idea, @SamuelMarks, can you try it?

**Comment by SamuelMarks** [comment]
@JonLiu1993 Hmm, I'm not sure if that will work. How do I pin it to a certain version?

There's no version number on the zip archive, and I can't see a way—correct me if you find one—for specifying `tinyfiledialogs-current.zip` at said hash.

**Review by JackBoosY — APPROVED** [review]

**Pull request merged**